[WPE] Provide a fallback WPEKeymap in new API's WPEDisplay class

lauromoura opened the pull request:
#### 60c2b80547ad004ccb91f1b211b2ce45500b92a7
<pre>
[WPE] Provide a fallback WPEKeymap in new API&apos;s WPEDisplay class
<a href="https://bugs.webkit.org/show_bug.cgi?id=275471">https://bugs.webkit.org/show_bug.cgi?id=275471</a>

Reviewed by Carlos Garcia Campos.

Allow some WebDriver tests issuing proper keyboard events instead of
failing early when using the headless display.

* Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
(wpe_display_get_keymap):

Canonical link: <a href="https://commits.webkit.org/280369@main">https://commits.webkit.org/280369@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d920f1a00afa2feee86deef09260c4ae2dbbc40b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56337 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35663 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8809 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/59944 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6773 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43285 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6967 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45631 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4726 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58366 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33543 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48608 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26494 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30322 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/5777 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52320 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6210 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61628 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/245 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6338 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52896 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/245 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48674 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/52779 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12491 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/222 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31490 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32576 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33659 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32323 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/headless/WPEDisplayHeadless.cpp:
```diff
@@ -69,6 +71,22 @@ static WPEBufferDMABufFormats* wpeDisplayHeadlessGetPreferredDMABufFormats(WPEDi
+    static std::once_flag onceFlag;
+    std::call_once(onceFlag, [display]() {
+        auto* priv = WPE_DISPLAY_HEADLESS(display)->priv;
+        priv->m_keymap = adoptGRef(wpe_keymap_xkb_new());
```
I wouldn't say this is a dummy keymap, it's a fallback. Since XKB is a common dependency, maybe we can use it always as fallback in `wpe_display_get_keymap`, and document that if platform doesn't implement get_keymap, a default one is created for a standard US 105 keyboard. We don't need call_once for this, this is always expectefd to be called from the main thread only.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/headless/WPEDisplayHeadless.cpp:
```diff
@@ -69,6 +71,22 @@ static WPEBufferDMABufFormats* wpeDisplayHeadlessGetPreferredDMABufFormats(WPEDi
+    if (!priv->m_keymap) {
+        g_set_error_literal(error, WPE_DISPLAY_ERROR, WPE_DISPLAY_ERROR_NOT_SUPPORTED, "Operation not supported");
+        return nullptr;
+    }
```
This can't happen, `wpe_keymap_xkb_new` always returns a new object.

lauromoura committed:
[WPE] Provide a fallback WPEKeymap in new API's WPEDisplay class
https://bugs.webkit.org/show_bug.cgi?id=275471

Reviewed by Carlos Garcia Campos.

Allow some WebDriver tests issuing proper keyboard events instead of
failing early when using the headless display.

* Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
(wpe_display_get_keymap):

Canonical link: https://commits.webkit.org/280369@main
